File: internship_backend_apr25/jury_solutions/python/e.py
import sys
from collections import defaultdict

class Program:
    @staticmethod
    def main():
        Program.solve()

    @staticmethod
    def solve():
        input_data = sys.stdin.read().splitlines()
        clusters_count, limit, q = map(int, input_data[0].split())
        servers_init = list(map(int, input_data[1].split()))

        clusters = [None] * clusters_count
        clusters_limits = [0] * clusters_count
        clusters_queries = [FixedQueue() for _ in range(clusters_count)]

        for i in range(clusters_count):
            servers_count = servers_init[i]
            if servers_count != 0:
                clusters[i] = CartesianNode()

            for j in range(1, servers_count):
                clusters[i] = CartesianNode.add(clusters[i], j)

        for t in range(q):
            line = input_data[t + 2].split()
            type_query = line[0][0]
            cluster = int(line[1]) - 1
            server = int(line[2])
            s = line[3]

            clusters_limits[cluster] += 1
            clusters_queries[cluster].add(server)
            current_server = clusters[cluster][server]

            if type_query == '+':
                current_server.server.add_string(s)
            elif type_query == 'p':
                current_server.server = current_server.server.add_prefix(s)
            elif type_query == 'c':
                print(current_server.server.count_prefix(s))

            if clusters_limits[cluster] == limit:
                clusters_limits[cluster] = 0
                next_cluster = (cluster + 1) % clusters_count
                from_index = min(clusters_queries[cluster].first, clusters_queries[cluster].last) - 1
                to_index = max(clusters_queries[cluster].first, clusters_queries[cluster].last)
                
                left, middle, right = None, None, None
                
                splitted = CartesianNode.split(clusters[cluster], to_index)
                left, right = splitted

                splitted = CartesianNode.split(left, from_index)
                left, middle = splitted

                clusters[cluster] = CartesianNode.merge(left, right)

                if clusters[next_cluster] is None:
                    clusters[next_cluster] = middle
                else:
                    clusters[next_cluster] = CartesianNode.insert(clusters[next_cluster], clusters[next_cluster].count // 2, middle)

class TrieNode:
    def __init__(self):
        self.next = {}
        self.count = 0

    def add_string(self, s, to_add=1):
        current_node = self
        current_node.count += to_add

        for c in s:
            if c not in current_node.next:
                current_node.next[c] = TrieNode()
            current_node = current_node.next[c]
            current_node.count += to_add

        return current_node

    def count_prefix(self, prefix):
        current_node = self

        for c in prefix:
            if c not in current_node.next:
                return 0
            current_node = current_node.next[c]

        return current_node.count

    def add_prefix(self, prefix):
        if self.count == 0:
            return self

        new_root = TrieNode()
        leaf = new_root.add_string(prefix, self.count)

        for i in self.next.items():
            leaf.next[i[0]] = i[1]

        return new_root

class CartesianNode:
    rand = __import__('random').Random()

    def __init__(self):
        self.height = CartesianNode.rand.randint(0, 1000000)
        self.count = 1
        self.server = TrieNode()
        self.left = None
        self.right = None

    @staticmethod
    def get_count(node):
        return 0 if node is None else node.count

    @staticmethod
    def update_count(node):
        if node is not None:
            node.count = CartesianNode.get_count(node.left) + CartesianNode.get_count(node.right) + 1

    @staticmethod
    def merge(left, right):
        if left is None:
            return right
        if right is None:
            return left

        if left.height > right.height:
            left.right = CartesianNode.merge(left.right, right)
            CartesianNode.update_count(left)
            return left
        else:
            right.left = CartesianNode.merge(left, right.left)
            CartesianNode.update_count(right)
            return right

    @staticmethod
    def split(node, key, add=0):
        if node is None:
            return None, None

        current_key = add + CartesianNode.get_count(node.left)
        if key <= current_key:
            splitted = CartesianNode.split(node.left, key, add)
            new_left, new_right = splitted
            node.left = new_right
            CartesianNode.update_count(node)
            return new_left, node
        else:
            splitted = CartesianNode.split(node.right, key, add + 1 + CartesianNode.get_count(node.left))
            new_left, new_right = splitted
            node.right = new_left
            CartesianNode.update_count(node)
            return node, new_right

    @staticmethod
    def insert(node, position, insert):
        splitted = CartesianNode.split(node, position)
        left, right = splitted
        res = CartesianNode.merge(left, insert)
        return CartesianNode.merge(res, right)

    @staticmethod
    def add(node, position):
        return CartesianNode.insert(node, position, CartesianNode())

    def __getitem__(self, index):
        count = CartesianNode.get_count(self.left) + 1
        if count == index:
            return self
        return self.right[index - count] if count < index else self.left[index]

class FixedQueue:
    def __init__(self):
        self.first = 0
        self.last = 0

    def add(self, index):
        self.first = self.last
        self.last = index


Program.main()

File: internship_backend_apr25/jury_solutions/cs/e.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;


class Program
{
    static void Main()
    {
        Solve();
    }

    public static void Solve()
    {
        var input = Console.ReadLine().Split();
        var clustersCount = int.Parse(input[0]);
        var limit = int.Parse(input[1]);
        var q = int.Parse(input[2]);
        
        var clusters = new List<CartesianNode>(clustersCount);
        var clustersLimits = new List<int>(clustersCount);
        var clustersQueries = new List<FixedQueue>(clustersCount);
        var serversInit = Console.ReadLine().Split().Select(int.Parse).ToArray();
        
        
        for (int i = 0; i < clustersCount; i++)
        {
            clustersQueries.Add(new FixedQueue());
            clustersLimits.Add(0);
            clusters.Add(null);

            var serversCount = serversInit[i];
            if (serversCount != 0)
                clusters[i] = new CartesianNode();

            for (int j = 1; j < serversCount; j++) clusters[i] = CartesianNode.Add(clusters[i], j);
        }
        
        for (int t = 0; t < q; t++)
        {
            var line = Console.ReadLine().Split();
            var type = line[0][0];
            var cluster = int.Parse(line[1]) - 1;
            var server = int.Parse(line[2]);
            var s = line[3];

            clustersLimits[cluster] += 1;
            clustersQueries[cluster].Add(server);
            var currentServer = clusters[cluster][server];

            switch (type)
            {
                case '+':
                    currentServer.Server.AddString(s);
                    break;
                case 'p':
                    currentServer.Server = currentServer.Server.AddPrefix(s);
                    break;
                case 'c':
                    Console.WriteLine(currentServer.Server.CountPrefix(s));
                    break;
            }

            if (clustersLimits[cluster] == limit)
            {
                clustersLimits[cluster] = 0;
                var nextCluster = (cluster + 1) % clustersCount;
                var from = Math.Min(clustersQueries[cluster].First, clustersQueries[cluster].Last) - 1;
                var to = Math.Max(clustersQueries[cluster].First, clustersQueries[cluster].Last);
                CartesianNode left, middle, right;
                
                var splitted = CartesianNode.Split(clusters[cluster], to);
                left = splitted.Item1;
                right = splitted.Item2;
                
                splitted = CartesianNode.Split(left, from);
                left = splitted.Item1;
                middle = splitted.Item2;
                
                clusters[cluster] = CartesianNode.Merge(left, right);

                if (clusters[nextCluster] == null)
                    clusters[nextCluster] = middle;
                else
                    clusters[nextCluster] =
                        CartesianNode.Insert(clusters[nextCluster], clusters[nextCluster].Count / 2, middle);
            }
        }
    }

    class TrieNode
    {
        Dictionary<char, TrieNode> Next = new Dictionary<char, TrieNode>();
        public long Count { get; set; } = 0;

        public TrieNode AddString(string s, long toAdd = 1)
        {
            var currentNode = this;
            currentNode.Count += toAdd;

            foreach (var c in s)
            {
                if (!currentNode.Next.TryGetValue(c, out var value))
                {
                    value = new TrieNode();
                    currentNode.Next[c] = value;
                }
                currentNode = value;
                currentNode.Count += toAdd;
            }
            
            return currentNode;
        }
        
        public long CountPrefix(string prefix)
        {
            var currentNode = this;

            foreach (var c in prefix)
            {
                if (!currentNode.Next.TryGetValue(c, out var value))
                    return 0;
                
                currentNode = value;
            }
            
            return currentNode.Count;
        }
        
        public TrieNode AddPrefix(string prefix)
        {
            if (Count == 0)
                return this;
            
            var newRoot = new TrieNode();
            TrieNode leaf = newRoot.AddString(prefix, Count);

            foreach (var i in Next) leaf.Next[i.Key] = i.Value;
            
            return newRoot;
        }
    }

    class CartesianNode
    {
        public long Height = rand.Next();
        public int Count = 1;
        public TrieNode Server = new TrieNode();
        private CartesianNode Left, Right;

        private static Random rand = new Random();
        
        public static int GetCount(CartesianNode node) => node == null ? 0 : node.Count;

        public static void UpdateCount(CartesianNode node)
        {
            if (node != null)
                node.Count = GetCount(node.Left) + GetCount(node.Right) + 1;
        }

        public static CartesianNode Merge(CartesianNode left, CartesianNode right)
        {
            if (left == null) return right;
            if (right == null) return left;

            if (left.Height > right.Height)
            {
                left.Right = Merge(left.Right, right);
                UpdateCount(left);
                return left;
            }
            else
            {
                right.Left = Merge(left, right.Left);
                UpdateCount(right);
                return right;
            }
        }

        public static Tuple<CartesianNode, CartesianNode> Split(CartesianNode node, int key, int add = 0)
        {
            if (node == null) return Tuple.Create<CartesianNode, CartesianNode>(null, null);

            var currentKey = add + GetCount(node.Left);
            if (key <= currentKey)
            {
                var splitted = Split(node.Left, key, add);
                var newLeft = splitted.Item1;
                var newRight = splitted.Item2;
                node.Left = newRight;
                UpdateCount(node);
                return Tuple.Create<CartesianNode, CartesianNode>(newLeft, node);
            }
            else
            {
                var splitted = Split(node.Right, key, add + 1 + GetCount(node.Left));
                var newLeft = splitted.Item1;
                var newRight = splitted.Item2;
                node.Right = newLeft;
                UpdateCount(node);
                return Tuple.Create<CartesianNode, CartesianNode>(node, newRight);
            }
        }

        public static CartesianNode Insert(CartesianNode node, int position, CartesianNode insert)
        {
            var splitted = Split(node, position);
            var left = splitted.Item1;
            var right = splitted.Item2;
            var res = Merge(left, insert);
            return Merge(res, right);
        }

        public static CartesianNode Add(CartesianNode node, int position) =>
            Insert(node, position, new CartesianNode());

        public CartesianNode this[int index]
        {
            get
            {
                var count = GetCount(Left) + 1;
                if (count == index)
                    return this;
                return count < index ? Right[index - count] : Left[index];
            }
        }
    }


    class FixedQueue
    {
        public int First => first;
        public int Last => last;
        private int first, last;

        public void Add(int index)
        {
            first = last;
            last = index;
        }
    }
}
